#**Credit Card Fraud Detection**

**Anonymized credit card transactions labeled as fraudulent or genuine**

## **About Dataset**

### **Context**

It is important that credit card companies are able to recognize fraudulent credit card transactions so that customers are not charged for items that they did not purchase.

### **Content**

The dataset contains transactions made by credit cards in September 2013 by European cardholders.
This dataset presents transactions that occurred in two days, where we have 492 frauds out of 284,807 transactions. The dataset is highly unbalanced, the positive class (frauds) account for 0.172% of all transactions.

It contains only numerical input variables which are the result of a PCA transformation. Unfortunately, due to confidentiality issues, we cannot provide the original features and more background information about the data. Features V1, V2, … V28 are the principal components obtained with PCA, the only features which have not been transformed with PCA are 'Time' and 'Amount'. Feature 'Time' contains the seconds elapsed between each transaction and the first transaction in the dataset. The feature 'Amount' is the transaction Amount, this feature can be used for example-dependant cost-sensitive learning. Feature 'Class' is the response variable and it takes value 1 in case of fraud and 0 otherwise.

Given the class imbalance ratio, we recommend measuring the accuracy using the Area Under the Precision-Recall Curve (AUPRC). Confusion matrix accuracy is not meaningful for unbalanced classification.

### **Update (03/05/2021)**

A simulator for transaction data has been released as part of the practical handbook on Machine Learning for Credit Card Fraud Detection - https://fraud-detection-handbook.github.io/fraud-detection-handbook/Chapter_3_GettingStarted/SimulatedDataset.html. We invite all practitioners interested in fraud detection datasets to also check out this data simulator, and the methodologies for credit card fraud detection presented in the book.

### **Acknowledgements**

The dataset has been collected and analysed during a research collaboration of Worldline and the Machine Learning Group (http://mlg.ulb.ac.be) of ULB (Université Libre de Bruxelles) on big data mining and fraud detection.
More details on current and past projects on related topics are available on https://www.researchgate.net/project/Fraud-detection-5 and the page of the DefeatFraud project

**Please cite the following works:**

Andrea Dal Pozzolo, Olivier Caelen, Reid A. Johnson and Gianluca Bontempi. Calibrating Probability with Undersampling for Unbalanced Classification. In Symposium on Computational Intelligence and Data Mining (CIDM), IEEE, 2015

Dal Pozzolo, Andrea; Caelen, Olivier; Le Borgne, Yann-Ael; Waterschoot, Serge; Bontempi, Gianluca. Learned lessons in credit card fraud detection from a practitioner perspective, Expert systems with applications,41,10,4915-4928,2014, Pergamon

Dal Pozzolo, Andrea; Boracchi, Giacomo; Caelen, Olivier; Alippi, Cesare; Bontempi, Gianluca. Credit card fraud detection: a realistic modeling and a novel learning strategy, IEEE transactions on neural networks and learning systems,29,8,3784-3797,2018,IEEE

Dal Pozzolo, Andrea Adaptive Machine learning for credit card fraud detection ULB MLG PhD thesis (supervised by G. Bontempi)

Carcillo, Fabrizio; Dal Pozzolo, Andrea; Le Borgne, Yann-Aël; Caelen, Olivier; Mazzer, Yannis; Bontempi, Gianluca. Scarff: a scalable framework for streaming credit card fraud detection with Spark, Information fusion,41, 182-194,2018,Elsevier

Carcillo, Fabrizio; Le Borgne, Yann-Aël; Caelen, Olivier; Bontempi, Gianluca. Streaming active learning strategies for real-life credit card fraud detection: assessment and visualization, International Journal of Data Science and Analytics, 5,4,285-300,2018,Springer International Publishing

Bertrand Lebichot, Yann-Aël Le Borgne, Liyun He, Frederic Oblé, Gianluca Bontempi Deep-Learning Domain Adaptation Techniques for Credit Cards Fraud Detection, INNSBDDL 2019: Recent Advances in Big Data and Deep Learning, pp 78-88, 2019

Fabrizio Carcillo, Yann-Aël Le Borgne, Olivier Caelen, Frederic Oblé, Gianluca Bontempi Combining Unsupervised and Supervised Learning in Credit Card Fraud Detection Information Sciences, 2019

Yann-Aël Le Borgne, Gianluca Bontempi Reproducible machine Learning for Credit Card Fraud Detection - Practical Handbook

Bertrand Lebichot, Gianmarco Paldino, Wissam Siblini, Liyun He, Frederic Oblé, Gianluca Bontempi Incremental learning strategies for credit cards fraud detection, IInternational Journal of Data Science and Analytics

# **Installing require Tools**

In [ ]:
pip install gradio

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 57.5/57.5 MB 12.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 320.6/320.6 kB 20.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 94.8/94.8 kB 7.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 11.3/11.3 MB 89.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 73.2/73.2 kB 4.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 62.3/62.3 kB 4.4 MB/s eta 0:00:00
  Attempting uninstall: markupsafe
    Found existing installation: MarkupSafe 3.0.2
    Uninstalling MarkupSafe-3.0.2:
      Successfully uninstalled MarkupSafe-3.0.2


#**Import the Libraries**

In [ ]:
## import some basic libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
import gradio as gr

#**Data Collecction and Preprocessing**

In [ ]:
# Loading the dataset into a pandas DataFrame
creditcard_data = pd.read_csv('creditcard.csv')  # Use this to identify issues

In [ ]:
# Display the first few rows
print("First 5 rows of the dataset:")
creditcard_data.head()


First 5 rows of the dataset:


Time        V1        V2        V3        V4        V5        V6        V7  \
0     0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1     0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2     1 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3     1 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4     2 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62    0.0  
1  0.125895 -0.008983  0.014724    2.69    0.0  
2 -0.139097 -0.055353 -0.059752  378.66    0.0  
3 -0.221929  0.062723  0.061458  123.50    0.0  
4  0.502292  0.219422  0.215153   69.99    0.0  

[5 rows x 31 columns]

In [ ]:
# Display the last few rows
print("Last 5 rows of the dataset:")
creditcard_data.tail()


Last 5 rows of the dataset:


Time        V1        V2        V3        V4        V5        V6  \
53566  46047 -2.885416 -2.647887  1.903441  3.726885  3.800798 -1.644480   
53567  46048  1.061410  0.045830  0.970401  1.102883 -0.490685  0.219968   
53568  46048 -0.202740  0.295074  1.924698  1.009344 -0.235931 -0.272232   
53569  46048 -0.730116  0.796820  2.122484 -1.318082  0.019762 -0.857857   
53570  46050  0.791672 -0.522991  0.036747  1.204803 -0.262390 -0.056153   

             V7        V8        V9  ...       V21       V22       V23  \
53566 -1.924198  0.707696 -1.070468  ...  0.284893 -0.268122  0.977417   
53567 -0.424285  0.235960  0.016498  ...  0.033158  0.067127  0.053801   
53568  0.261015 -0.162146 -0.583849  ...  0.321731  1.056540  0.093902   
53569  0.975752 -0.399896  0.359189  ... -0.149736  0.093479 -0.056586   
53570  0.263614 -0.130209  0.125423  ...  0.013131 -0.000000       NaN   

            V24       V25       V26       V27       V28  Amount  Class  
53566  0.136853 -1.144367 -0.535616  0.098494  0.305740   19.02    0.0  
53567 -0.010221  0.209877 -0.484136  0.057583  0.028018   28.70    0.0  
53568  0.608566 -0.931546  0.466156  0.016695 -0.024568   54.81    0.0  
53569  0.650087 -0.473889  0.662115  0.060486 -0.208536    4.53    0.0  
53570       NaN       NaN       NaN       NaN       NaN     NaN    NaN  

[5 rows x 31 columns]

In [ ]:
# Checking the shape of the dataset
creditcard_data.shape    # Outputs the number of rows and columns in the dataset.

(53571, 31)

In [ ]:
# Statistical description of the dataset
print("\nStatistical Description of the Dataset:")
creditcard_data.describe()


Statistical Description of the Dataset:


Time            V1            V2            V3            V4  \
count  53571.000000  53571.000000  53571.000000  53571.000000  53571.000000   
mean   30004.864404     -0.243182      0.000586      0.691588      0.179664   
std    13302.545313      1.879598      1.639428      1.483708      1.393856   
min        0.000000    -56.407510    -72.715728    -32.965346     -5.172595   
25%    23663.500000     -0.994263     -0.573628      0.212411     -0.724005   
50%    34251.000000     -0.249472      0.078153      0.790661      0.189575   
75%    40127.000000      1.154647      0.732455      1.424778      1.061933   
max    46050.000000      1.960497     18.183626      4.101716     16.491217   

                 V5            V6            V7            V8            V9  \
count  53571.000000  53571.000000  53571.000000  53571.000000  53571.000000   
mean      -0.257872      0.102917     -0.119311      0.052872      0.094530   
std        1.411097      1.313136      1.271570      1.214929      1.201748   
min      -42.147898    -26.160506    -26.548144    -41.484823     -9.283925   
25%       -0.873897     -0.636806     -0.606014     -0.147112     -0.630279   
50%       -0.290113     -0.152437     -0.075304      0.058859     -0.012999   
75%        0.281788      0.493733      0.425496      0.333521      0.780864   
max       34.801666     22.529298     36.677268     20.007208     10.392889   

       ...           V21           V22           V23           V24  \
count  ...  53571.000000  53571.000000  53570.000000  53570.000000   
mean   ...     -0.028932     -0.106714     -0.039090      0.007830   
std    ...      0.731807      0.636939      0.593273      0.596080   
min    ...    -20.262054     -8.593642    -26.751119     -2.836627   
25%    ...     -0.230582     -0.528974     -0.179410     -0.322721   
50%    ...     -0.066124     -0.082370     -0.051736      0.062416   
75%    ...      0.110098      0.307953      0.079102      0.401967   
max    ...     22.614889      5.805795     17.297845      4.014444   

                V25           V26           V27           V28        Amount  \
count  53570.000000  53570.000000  53570.000000  53570.000000  53570.000000   
mean       0.135096      0.020158      0.003957      0.003972     94.260870   
std        0.438590      0.500092      0.388427      0.329538    252.290845   
min       -7.495741     -1.577118     -8.567638     -9.617915      0.000000   
25%       -0.128328     -0.330573     -0.063648     -0.006589      7.680000   
50%        0.174351     -0.073699      0.008850      0.022405     25.495000   
75%        0.421204      0.298577      0.083071      0.076498     86.867500   
max        5.525093      3.517346     11.135740     33.847808  12910.930000   

              Class  
count  53570.000000  
mean       0.002856  
std        0.053366  
min        0.000000  
25%        0.000000  
50%        0.000000  
75%        0.000000  
max        1.000000  

[8 rows x 31 columns]

In [ ]:
# Check for missing values
missing_values = creditcard_data.isnull().sum()
print("\nMissing values in each column:")
print(missing_values)


Missing values in each column:
Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       1
V24       1
V25       1
V26       1
V27       1
V28       1
Amount    1
Class     1
dtype: int64


In [ ]:
# Geerating some Information about the dataset
creditcard_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 53571 entries, 0 to 53570
Data columns (total 31 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Time    53571 non-null  int64  
 1   V1      53571 non-null  float64
 2   V2      53571 non-null  float64
 3   V3      53571 non-null  float64
 4   V4      53571 non-null  float64
 5   V5      53571 non-null  float64
 6   V6      53571 non-null  float64
 7   V7      53571 non-null  float64
 8   V8      53571 non-null  float64
 9   V9      53571 non-null  float64
 10  V10     53571 non-null  float64
 11  V11     53571 non-null  float64
 12  V12     53571 non-null  float64
 13  V13     53571 non-null  float64
 14  V14     53571 non-null  float64
 15  V15     53571 non-null  float64
 16  V16     53571 non-null  float64
 17  V17     53571 non-null  float64
 18  V18     53571 non-null  float64
 19  V19     53571 non-null  float64
 20  V20     53571 non-null  float64
 21  V21     53571 non-null  float64
 22

In [ ]:
# Distribution of Legit transactions & Fraudulent Transactions
creditcard_data['Class'].value_counts()

Class
0.0    53417
1.0      153
Name: count, dtype: int64

## This Dataset is highly Unbalanced

**0 --->  Normal Transaction**

**1 ---> Fraudulent Transaction**


In [ ]:
# Separating the data for analysis
legit = creditcard_data[creditcard_data.Class == 0]
fraud = creditcard_data[creditcard_data.Class == 1]

In [ ]:
print(legit.shape)
print(fraud.shape)

(53417, 31)
(153, 31)


In [ ]:
# Statistical Measures of the data
legit.Amount.describe()

count    53417.000000
mean        94.251260
std        252.353693
min          0.000000
25%          7.680000
50%         25.520000
75%         86.650000
max      12910.930000
Name: Amount, dtype: float64

In [ ]:
fraud.Amount.describe()

count     153.000000
mean       97.616013
std       230.026856
min         0.000000
25%         1.000000
50%         7.610000
75%        99.990000
max      1809.680000
Name: Amount, dtype: float64

In [ ]:
# Compare the values for both transaction
creditcard_data.groupby('Class').mean()

Time        V1        V2         V3        V4        V5  \
Class                                                                   
0.0    30011.81811 -0.222505 -0.014659   0.722511  0.163565 -0.242775   
1.0    27472.24183 -7.468948  5.326389 -10.100308  5.793812 -5.528864   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0.0    0.109604 -0.097131  0.042665  0.104805  ...  0.046251 -0.031568   
1.0   -2.230554 -7.865700  3.617647 -3.492917  ...  0.455720  0.891045   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0.0   -0.106396 -0.038408  0.008094  0.134794  0.019825  0.002289  0.003902   
1.0   -0.218399 -0.277419 -0.084305  0.240593  0.136258  0.586389  0.028185   

          Amount  
Class             
0.0    94.251260  
1.0    97.616013  

[2 rows x 30 columns]

## **Under-Sampling**


Build a sample dataset containing similar distribution of Legit Transactions and Fraudulent Transactions



**Number of Normal Transactions --->53417**

**Number of Fraudulent Transactions ---> 153**



Picking randomly 153-data from the Legit dataset to equal to the number of Fraud dataset


**NB:**

But this method isn't the method for this data because amount have been lost

In [ ]:
legit_sample = legit.sample(n=153)

In [ ]:
# Concatenating the two DataFrames
new_dataset = pd.concat([legit_sample, fraud], axis=0)

In [ ]:
# Display the first few rows
print("First 5 rows of the dataset:")
new_dataset.head()

First 5 rows of the dataset:


Time        V1        V2        V3        V4        V5        V6  \
10031  15124 -2.149789  0.972363 -0.233447 -0.887317  1.246592  3.785172   
39914  40019 -0.621262 -1.274220  1.169739 -2.517339 -3.105711  0.357250   
388      283 -0.529996  0.766554  1.759393 -1.160074 -0.501040 -1.404513   
20        16  0.694885 -1.361819  1.029221  0.834159 -1.191209  1.309109   
9004   12493 -1.767308  1.153162 -0.480875 -0.492061  2.841513  3.236165   

             V7        V8        V9  ...       V21       V22       V23  \
10031 -0.606391  1.730288  0.819285  ... -0.256135 -0.725039 -0.034730   
39914  1.536640 -0.529384  1.228491  ... -0.466212 -0.257954  0.271784   
388    0.679279 -0.242594  0.520868  ... -0.163031 -0.219408  0.016959   
20    -0.878586  0.445290 -0.446196  ... -0.295583 -0.571955 -0.050881   
9004  -0.117862  0.471026  1.703544  ... -0.608585 -1.052747 -0.065445   

            V24       V25       V26       V27       V28  Amount  Class  
10031  0.954412  0.384467  0.241489  0.066946 -0.005989  119.99    0.0  
39914  0.385000 -0.440176 -0.872530  0.193489 -0.153349  473.28    0.0  
388    0.934128 -0.327383  0.668479  0.114264 -0.091385    2.31    0.0  
20    -0.304215  0.072001 -0.422234  0.086553  0.063499  231.71    0.0  
9004   0.899702 -0.324471  0.103059 -0.020932  0.364630    9.77    0.0  

[5 rows x 31 columns]

In [ ]:
# Checking of Legit transactions & Fraudulent Transactions
new_dataset['Class'].value_counts()

Class
0.0    153
1.0    153
Name: count, dtype: int64

In [ ]:
new_dataset.groupby('Class').mean()

Time        V1        V2         V3        V4        V5  \
Class                                                                    
0.0    30217.294118 -0.661859  0.006232   0.407257  0.321447 -0.421282   
1.0    27472.241830 -7.468948  5.326389 -10.100308  5.793812 -5.528864   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0.0    0.206442 -0.156453  0.204143  0.154880  ... -0.062818 -0.074177   
1.0   -2.230554 -7.865700  3.617647 -3.492917  ...  0.455720  0.891045   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0.0   -0.088759  0.006026  0.039516  0.106809  0.067577  0.029473  0.009340   
1.0   -0.218399 -0.277419 -0.084305  0.240593  0.136258  0.586389  0.028185   

           Amount  
Class              
0.0    122.978693  
1.0     97.616013  

[2 rows x 30 columns]

# **Splitting the data into Features & Target**


In [ ]:
X = new_dataset.drop(columns='Class', axis=1)
y= new_dataset['Class']

In [ ]:
print("Features (X):")
print(X.head())
print("\nTarget (y):")
print(y.head())

Features (X):
        Time        V1        V2        V3        V4        V5        V6  \
10031  15124 -2.149789  0.972363 -0.233447 -0.887317  1.246592  3.785172   
39914  40019 -0.621262 -1.274220  1.169739 -2.517339 -3.105711  0.357250   
388      283 -0.529996  0.766554  1.759393 -1.160074 -0.501040 -1.404513   
20        16  0.694885 -1.361819  1.029221  0.834159 -1.191209  1.309109   
9004   12493 -1.767308  1.153162 -0.480875 -0.492061  2.841513  3.236165   

             V7        V8        V9  ...       V20       V21       V22  \
10031 -0.606391  1.730288  0.819285  ...  0.071599 -0.256135 -0.725039   
39914  1.536640 -0.529384  1.228491  ... -0.307981 -0.466212 -0.257954   
388    0.679279 -0.242594  0.520868  ...  0.138236 -0.163031 -0.219408   
20    -0.878586  0.445290 -0.446196  ... -0.138334 -0.295583 -0.571955   
9004  -0.117862  0.471026  1.703544  ...  0.404533 -0.608585 -1.052747   

            V23       V24       V25       V26       V27       V28  Amount  
10031 -0

# **Splitting the dataset into Training and Test sets**

In [ ]:
# Splitting the dataset into Training and Test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, stratify=y, random_state=3)
print("\nDataset split completed:")
print(f"Total samples: {X.shape[0]}, Training samples: {X_train.shape[0]}, Test samples: {X_test.shape[0]}")


Dataset split completed:
Total samples: 306, Training samples: 229, Test samples: 77


In [ ]:
# checking the number of Test and Train dataset
print(X.shape, X_train.shape, X_test.shape)

(306, 30) (229, 30) (77, 30)


# **Model Training**

In [ ]:
# Initialize the model
logReg_model = LogisticRegression()

In [ ]:
# Training the LogisticRegression model with train data
logReg_model.fit(X_train, y_train)

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

#**Model Evaluation**

In [ ]:
# Calculate accuracy on the training data
X_train_pred = logReg_model.predict(X_train)
train_data_accuracy = accuracy_score(X_train_pred, y_train)
print('Accuracy on training data : ', train_data_accuracy)


Accuracy on training data :  0.9606986899563319


In [ ]:
# Generate and display the confusion matrix on the training data
# The confusion matrix shows the counts of True Positives, True Negatives, False Positives, and False Negatives
conf_matrix = confusion_matrix(y_train, X_train_pred)
print("\nConfusion Matrix:")
print(conf_matrix)


Confusion Matrix:
[[111   3]
 [  6 109]]


In [ ]:
# Generate and display the classification report
# The classification report includes precision, recall, F1-score, and support for each class
class_report = classification_report(y_train, X_train_pred, target_names=["Normal Transaction (0)", "Fraudulent Transaction (1)"])
print("\nClassification Report:")
print(class_report)


Classification Report:
                            precision    recall  f1-score   support

    Normal Transaction (0)       0.95      0.97      0.96       114
Fraudulent Transaction (1)       0.97      0.95      0.96       115

                  accuracy                           0.96       229
                 macro avg       0.96      0.96      0.96       229
              weighted avg       0.96      0.96      0.96       229



In [ ]:
# accuracy on the test data
X_test_pred = logReg_model.predict(X_test)
test_data_accuracy = accuracy_score(X_test_pred, y_test)
print('Accuracy on test data : ', test_data_accuracy)

Accuracy on test data :  0.9090909090909091


In [ ]:
# Generate and display the confusion matrix on the test data
# The confusion matrix shows the counts of True Positives, True Negatives, False Positives, and False Negatives
conf_matrix = confusion_matrix(y_test, X_test_pred)
print("\nConfusion Matrix:")
print(conf_matrix)


Confusion Matrix:
[[38  1]
 [ 6 32]]


In [ ]:
# Generate and display the classification report
# The classification report includes precision, recall, F1-score, and support for each class
class_report = classification_report(y_test, X_test_pred, target_names=["Normal Transaction (0)", "Fraudulent Transaction (1)"])
print("\nClassification Report:")
print(class_report)


Classification Report:
                            precision    recall  f1-score   support

    Normal Transaction (0)       0.86      0.97      0.92        39
Fraudulent Transaction (1)       0.97      0.84      0.90        38

                  accuracy                           0.91        77
                 macro avg       0.92      0.91      0.91        77
              weighted avg       0.92      0.91      0.91        77

